### ISSUE TITLE
Replacing parts of .msg lines

### ISSUE THREAD

**Issue by wipe2238** [phobos2077/sfall#274]
_NOTE: This is pretty long, but please don't read it as a request; it's more like a bunch of ideas how 
stuff could be implemented to make it interesting to modders._

db4652c reminded me of _tags_ system we had in FOnline. Converted to Fallout reality, and tweaked a bit it could become something like this some day:

##### dude_gender
`{1}{}{@dude_gender|Howdy pardner.|M'lady..@. What brings you here? @dude_gender||**tips fedora**@}`
```
(Male chars) Howdy pardner. What brings you here?
(Female chars) M'lady... What brings you here? **tips fedora**
```

##### dude_name
`{1}{}{I lost my memory! Who am I?}`
`{2}{}{Stop fooling around, @dude_name@. You're not in pre-war movie...}`

```
- I lost my memory! Who am I?!
- Stop fooling around, NovaRain. You're not in pre-war movie...}
```

##### npc_name
`{1}{}{I’m @npc_name@. Helped build the place. Now I’m sheriff.}`
```
I’m Marcus. Helped build the place. Now I’m sheriff.
```

##### Reporting unknown replacement (due to typo)
`{1}{}{I’m @npc_naem@. Helped build the place. Now I’m sheriff.}`
```
I’m [ERROR: unknown tag npc_naem]. Helped build the place. Now I’m sheriff.
```

TL;DR
- All replacements needs to start __and__ end with magic character (yes, i know `@` can be a problem, it's just example)
- Both prefix and suffix had to be present for replacing to kick in, which results in ignoring lines like `{1}{}{Send bug reports to someone@somewhere.com}`
- Another magic character (for example `|`) is added to extract "arguments" for given replacement
- Each replacement have its own unique name, so it's possible to add more of them in future
- Unknown replacements names returns string with warning

With all that in place, introducing on/off option is unneeded; chances that mod/vanilla fallouts uses text which collides with that are pretty low, i think, but could be wrong. It's also consistent, allows replacing with more than one word, and keeps a space to add new replacements in future.

---

Obviously, only first and last example is needed right now, other's might be added later, if ever. One thing i really liked in FOnline was that i could add .msg lines which used scripts to fill parts of text; helped a lot when some script hackery was involved. I don't know how realistic it is here, but converting this idea to Fallout modding could look like this:

`{1}{}{@script|gl_dialog_tags|gender@}`
`{1}{}{Helo @script|gl_dialog_tags|replace|gender|dude|dudette@!}`
`{1}{}{It's over @script|gl_dialog_tags|replace|power_level@!!!}`
This would call `replace` and `gender` procedures in `gl_dialog_tags` script, with arrays as argument.
```pascal
// gl_dialog_tags.ssl; some pseudocode here

procedure gender(variable args, variable cr)
begin
    // args = array arguments used in @script@, if any; can be empty, cannot be null
    // cr = critter saying given text; can be dude_obj; can be null if used from weird places

    if( critter_is_male( cr )) then
         return "male";
    else
         return "female";
end

procedure replace(variable args, variable cr)
begin
    // args = array arguments used in @script@, if any; can be empty, cannot be null
    // cr = critter saying given text; can be dude_obj; can be null if used from weird places

    if( array_size( args ) == 0 ) then
        return "[ERROR: missing tag]";

    // copy arg[0] to variable, remove it from array
    variable tag := array_pop( args, 0 );

    if( tag == "gender" ) then
    begin
        if( array_size( args ) != 2 )
            return "[ERROR: tag gender needs 2 arguments]";

        return args[get_critter_stat(cr,STAT_gender)];
    end
    else if( tag == "power_level" ) then
         return "" + get_critter_stat(cr,STAT_st) + "000";

    return "[ERROR: unknown tag " + tag + "]";
end
```

This could be **THE** way to implement .msg replacements - remember `@dude_name@`, etc? It wouldn't be even needed with `@script@`; modders could add tags or their own (or not), take examples from modders pack/NMA/etc. I'm using global script here as tags provided by example script are very generic (and available all the time), but it should work with regular scripts too, i think.

Tags provided by sfall could be only things which are impossible to get from scripts and stuff which should work even when scripts are loaded yet. Can't think of good examples here honestly :D but i'd leave a space for that, if that's needed some day.

Making `@script@` a central point of the feature would keep it alive; if script can read some data, it can be used in .msg - gvars as text, protos data, .ini entries, ingame hour, amount of rats killed, you name it. Let's be honest, nobody would add even a half of it to sfall. But having an option to do stuff in scripts there's no need to bother anyone; just do what you need in scripts and be done with it. Replacements being done 100% in .dll is interesting, but have no real future and place to grow.

**Comment by FakelsHub** [comment]
It is too difficult to add to sfall, it is easier to do it all in a script with its own macro-functions `nTagOption/nTagReply`.

**Comment by wipe2238** [comment]
Hehe, though so. But maybe at least use format which allows >1 word for each gender? That's a good start.

**Comment by FakelsHub** [comment]
Well if only to redone the current implementation to the form: `~Howdy pardner.^M'lady.~`

**Comment by FakelsHub** [comment]
`~` - can be used for national symbols. then I think `|` or `@`

**Comment by wipe2238** [comment]
I think it's better to leave `~` out of this, so it still can be used in dialogs. Same with `<`/`>`, i'm near sure they are used in terminal dialogs somewhere.

How about `^Howdy pardner.|Oh my princess!^`?

**Comment by FakelsHub** [comment]
No, the symbol `|` can be used.

**Comment by burner1024** [comment]
A bit late, but some notes:
- There's [tokenize](https://github.com/phobos2077/sfall/blob/5200be19c90560050efa3a7a3f168faa6e6c3d72/artifacts/scripting/headers/lib.strings.h#L180) already, allowing to embed gvars, protos data, .ini entries, whatever.
-
  ```
  {1}{}{@script|gl_dialog_tags|gender@}

  procedure gender(variable args, variable cr)
  begin
    if( critter_is_male( cr )) then
         return "male";
    else
         return "female";
  end
  ```
  Is not translation-friendly. It was an issue in original F2 as well, they just hardcoded some english strings into scripts.
- Again, from translator's point of view:
  - `{2}{}{Stop fooling around, @dude_name@. You're not in pre-war movie...}`
     says: okay... should probably just copy `@dude_name@` as it is...
   - `{1}{}{@dude_gender|Howdy pardner.|M'lady..@. What brings you here? @dude_gender||**tips fedora**@}`
     and
      `{1}{}{Helo @script|gl_dialog_tags|replace|gender|dude|dudette@!}`
    say: "I DARE YOU NOT TO MAKE A TYPO LITTLE MAN!"

The main draw of having tokens would be to move some logic from scripts to messages, hopefully improving readability of both. But I don't think it's wise to move too much. Scripts are for scripting. Messages are for text.

Regarding multi-word phrases, in my opinion it's such a marginal use case that I'd just bite the bullet and make an if/else in the code.
- Alternatively, I think it's possible to use a convention such as
  ```
  {1}{}{Howdy pardner. What brings you here?}
  {99991}{}{M'lady. What brings you here? **tips fedora**}
  ```
  and then pull a correct one with a macro.
- Yet alternatively, msg format itself could probably be extended to allow female strings embedded:
  ```
  {1}{}{Howdy pardner. What brings you here?} /# M'lady. What brings you here? **tips fedora** #/
  ```
  (But this would make translation process even more messy, so I don't like this idea, just including for completeness).

Ultimately, my point is that code is read many more times than written. MSG files aren't exactly code, but that applies to them too. So I would ask to prioritize simplicity over typing a few lines less.

**Comment by FakelsHub** [comment]
> Regarding multi-word phrases, in my opinion it's such a marginal use case that I'd just bite the bullet and make an if/else in the code.

I also believe that different texts for F/M need to be done in scripts, so initially my implementation was only for one word, to simplify the use of additional conditions only because of one "female" word.
And since asked for a multi-word version, please, because there is no significant difference in the implementation.
All other perversions with the text should be done in scripts.
